swim/join_sender.go

```diff
@@ -73,6 +77,10 @@ type joinSender struct {
 	node    *Node
 	timeout time.Duration
 
+	// bootstrapHostsMap is a map of unique hosts each containing a slice of
+	// the instances (hostsports) on that particular host.
+	bootstrapHostsMap map[string][]string
```

Comment:
What do you mean by "Bootstrap Selector-type-of-abstraction"? Are you referring to the hostProvider/discoverProvider?

See my comment here:
https://github.com/uber/ringpop-go/pull/59#discussion_r48941935

Right now, the grouping of hosts is a concern of the JoinSender. If we were to push this concern to the discoverProvider, we would need to push the entire selection algorithm (i.e. which nodes to contact first) there as well, no? Which doesn't seem right, because that algorithm is dependent on local information about the current node (e.g. do I want to contact peers that are on the same host as me, or not).